The dependency on `css-tree` causes bundling issues

So far my project has been working & bundling well with `nytimes/react-prosemirror`. Unfortunately, after switching to this package, at runtime (not in the browser but during the SSR stage) I get `Error: Cannot find module '../data/patch.json'`.

It’s very likely a bundling issue. My project is somewhat atypical in things like environment-agnostic SSR (ability to generate static HTML in browser context, not just in Node/Bun/Deno like most frameworks), support for pluggable runtime dependencies, etc., which definitely has made me sensitive to how libraries can be engineered in ways that can make them a breeze or a pain to bundle.

Looks like some other people encountered this with `css-tree`, too: https://github.com/sveltejs/svelte/issues/9288#issuecomment-2333516516. You can observe a chain of workarounds (none of which help me in particular, I’m using esbuild and not using Svelte) and in the end they removed the dependency on `css-tree` altogether, for a variety of reasons.

- I poked around their tickets and found some [other](https://github.com/sveltejs/svelte/issues/8670#issuecomment-1668822597) [minor](https://github.com/sveltejs/svelte/issues/8235#issuecomment-2035186524) [and major](https://github.com/sveltejs/svelte/issues/9065#issuecomment-1663561815)  [issues](https://github.com/sveltejs/svelte/issues/5876#issuecomment-1453750503) they had with `css-tree`. They apparently had to keep augmenting it with custom extensions until they figured it was easier to just have a full home-grown parser instead. I suspect you might run into these other issues with styling sooner or later, but I’m not sure what’s a more sustainable alternative.
- I poked around Svelte’s codebase, and it looks like the logic where they *used to* rely on `css-tree` up until last year now looks [like this](https://github.com/sveltejs/svelte/blob/main/packages/svelte/src/compiler/phases/1-parse/read/style.js)—a bit of dense parsing which honestly seems fit for splitting into a library (would be nice of them, at least). I’m not sure whether it’s something you would be interested in, of course—there’re different tradeoffs between maintaining a dependency on a third-party library vs. making more logic part of the codebase.

\*\*\*

Incidentally, there’s another mention in Svelte that brings up css-tree’s bundling: see the crossed-through text in [this comment](https://github.com/sveltejs/svelte/issues/6969#issuecomment-1407270452) and a hint at a possible solution of changing how it’s imported—I might try patching/forking RPM to see if it helps my case (and if so perhaps make a PR):

> The problem described above was solved in `css_tree_parse.ts` by not importing `fork` from the `css-tree` Node package, but importing it from the `css-tree` browser distribution bundle...
> 
> Changing this: `import { fork } from 'css-tree';`
> 
> to:
> 
> `import { fork } from '../../../../../node_modules/css-tree/dist/csstree.esm.js';`

--- a/.yarn/versions/22ef9f0f.yml
+++ b/.yarn/versions/22ef9f0f.yml
@@ -1,0 +1,2 @@
+releases:
+  "@handlewithcare/react-prosemirror": patch

--- a/package.json
+++ b/package.json
@@ -122,7 +122,6 @@
   },
   "dependencies": {
     "classnames": "^2.3.2",
-    "css-tree": "^3.1.0",
     "react-reconciler": "0.31.0"
   }
 }

--- a/src/props.ts
+++ b/src/props.ts
@@ -1,6 +1,30 @@
 import cx from "classnames";
 import { HTMLProps } from "react";
 
+let patched = false;
+
+function patchConsoleError() {
+  if (patched) return;
+  /* eslint-disable no-console */
+  const consoleError = console.error;
+  console.error = function (...data: unknown[]) {
+    const [message, prop, correction] = data;
+    if (
+      typeof message === "string" &&
+      message.startsWith(
+        "Warning: Invalid DOM property `%s`. Did you mean `%s`?"
+      ) &&
+      prop === "STYLE" &&
+      correction === "style"
+    ) {
+      return;
+    }
+    consoleError(...data);
+  };
+  patched = true;
+  /* eslint-enable no-console */
+}
+
 function mergeStyleProps(a: HTMLProps<HTMLElement>, b: HTMLProps<HTMLElement>) {
   if (!("STYLE" in a)) {
     if (!("STYLE" in b)) {
@@ -44,6 +68,7 @@ export function mergeReactProps(
 export function htmlAttrsToReactProps(
   attrs: Record<string, string>
 ): HTMLProps<HTMLElement> {
+  patchConsoleError();
   const props: Record<string, unknown> = {};
   for (const [attrName, attrValue] of Object.entries(attrs)) {
     switch (attrName.toLowerCase()) {

--- a/src/props.ts
+++ b/src/props.ts
@@ -1,41 +1,19 @@
 import cx from "classnames";
-import { generate, parse } from "css-tree";
 import { HTMLProps } from "react";
 
-export function kebabCaseToCamelCase(str: string) {
-  return str.replaceAll(/-[a-z]/g, (g) => g[1]?.toUpperCase() ?? "");
-}
-
-/**
- * Converts a CSS style string to an object
- * that can be passed to a React component's
- * `style` prop.
- */
-export function cssToStyles(css: string) {
-  const ast = parse(`* { ${css} }`);
-
-  if (ast.type !== "StyleSheet") return {};
-
-  const rule = ast.children.first;
-  if (rule?.type !== "Rule") return {};
-
-  const block = rule.block;
-  const styles: Record<string, string> = {};
-
-  for (const declaration of block.children) {
-    if (declaration.type !== "Declaration") continue;
-    const property = declaration.property;
-    const value =
-      declaration.value.type === "Raw"
-        ? declaration.value.value
-        : generate(declaration.value);
-    const camelCasePropertyName = property.startsWith("--")
-      ? property
-      : kebabCaseToCamelCase(property);
-    styles[camelCasePropertyName] = value;
+function mergeStyleProps(a: HTMLProps<HTMLElement>, b: HTMLProps<HTMLElement>) {
+  if (!("STYLE" in a)) {
+    if (!("STYLE" in b)) {
+      return undefined;
+    }
+    return b.STYLE as string;
   }
-
-  return styles;
+  if (!("STYLE" in b)) {
+    return a.STYLE as string;
+  }
+  return `${(a.STYLE as string).match(/;\s*$/) ? a.STYLE : `${a.STYLE}`} ${
+    b.STYLE
+  }`;
 }
 
 /**
@@ -51,6 +29,7 @@ export function mergeReactProps(
     ...a,
     ...b,
     className: cx(a.className, b.className),
+    STYLE: mergeStyleProps(a, b),
     style: {
       ...a.style,
       ...b.style,
@@ -73,7 +52,14 @@ export function htmlAttrsToReactProps(
         break;
       }
       case "style": {
-        props.style = cssToStyles(attrValue);
+        // HACK: React expects the `style` prop to be an
+        // object mapping from CSS property name to value.
+        // However, it will pass un-recognized props through
+        // to the underlying DOM element, and HTML attributes
+        // are case insensitive. So we use `STYLE` instead,
+        // which React doesn't intercept, but the DOM treats
+        // as `style`
+        props.STYLE = attrValue;
         break;
       }
       case "autocapitalize": {

